CCO-84: ccoctl upgrade process (#36273)
openshift/openshift-docs: modules/cco-ccoctl-upgrading.adoc

.Procedure

. Extract the list of `CredentialsRequest` objects from the {product-title} release image:
+
[source,terminal,subs="+quotes"]
----
$ oc adm release extract --credentials-requests --cloud=aws --to=__<path_to_directory_with_list_of_credentials_requests>__/credrequests quay.io/__<path_to>__/ocp-release:__<version>__
----

. Complete appropriate health checks. #elaborate#
//Test plan has a link to several health checks, but what specifically should the user run and what are they watching out for?

. Use the `ccoctl` tool to process all `CredentialsRequest` objects in the `credrequests` directory:
+
[source,terminal,subs="+quotes"]
----
$ ccoctl aws create-iam-roles --name __<name>__ --region=__<aws_region>__ --credentials-requests-dir=__<path_to_directory_with_list_of_credentials_requests>__/credrequests --identity-provider-arn arn:aws:iam::__<aws_account_id>__:oidc-provider/__<name>__-oidc.s3.__<aws_region>__.amazonaws.com
----
+
where:
+

> ** _<name>_ is the name used to tag any cloud resources that are created for tracking.

Reviewer: ```suggestion
** _<name>_ is the name used to tag any cloud resources that are created for tracking. For upgrades, this name field should be the same as was used before installation.
```
(otherwise, you end up with completely new AWS resources, which is unlikely to be desired)

** _<aws_account_id>_ is the AWS account ID.
** _<aws_region>_ is the AWS region in which cloud resources will be created.
+
[NOTE]
====
For AWS environments that use alternative IAM API endpoints, such as GovCloud, you must also specify your region with the `--region` parameter.
====
+
For each `CredentialsRequest` object, `ccoctl` creates an IAM role with a trust policy that is tied to the specified OIDC identity provider, and a permissions policy as defined in each `CredentialsRequest` object from the {product-title} release image.

> . Add any new permissions in the appropriate role and policy on AWS.

Reviewer: this already happened when you ran the command above. the IAM Roles should be up-to-date at this point.

//At this point in the test plan, it is stated that "there are 2 new permissions added for machine-api." How does the user figure this out? Would output from the previous step help?

> //Not really clear on the correct thing to tell the user to do on the AWS side, what is this process called? Any AWS docs we should refer them to?

Reviewer: there would only be new permission added if that specific release image being upgraded to added those new permissions in that release. in the common case, there are typically no new permissions added.

Author: ok, and per your previous comment it sounds like no additional action is required for those new permissions after they run `ccoctl aws create-iam-roles [...]`, correct?

Reviewer: correct

+

> It may take up to several minutes for the updated permissions to take effect.

Reviewer: I've seen it take a couple seconds, but never minutes. But if we have reason to state that it make take several minutes, then I guess we leave this more broad caveat in this sentence as-is.

. Complete appropriate health checks. #elaborate#
//Again, test plan has a link to several health checks, but what specifically should the user run and what are they watching out for?

.Verification

> //Does this make sense for the upgrade as well?

Reviewer: I don't know that viewing these files helps. These files are typically passed to the installer, and the installer will install/apply them. For the upgrade case, we simply need to apply all the files that end in "credentials.yaml".
So something like:
ls _<path_to_ccoctl_output_dir>_/manifests/*-credentials.yaml | xargs -I{} oc apply -f {}

What may trip someone up is if there is a new CredentialsRequest that wants the Secret placed in a namespace that does not yet exist (it would be created as part of the actual upgrade process). The above command will fail to create those Secrets because of the missing namespace, and you would need to `oc create namespace <name of missing namespace>` for each Secret that failed to apply because of non-existent namespace and re-run the above command.

At this point, the IAM Roles are update, any new IAM Roles have been created, and we've applied the Secrets for the new component(s) (if there were new CredentialsRequests in this specific release image).

Reviewer: On the upgrade case, we get no help. We have to do what the installer would do for us. Which is basically just taking those yaml files with Secret data in them, and just applying them once the control plane is up and running.
So for upgrade, we:
1) generate the IAM Roles (from the list of extracted CredentialsRequests)
2) generate the OpenShift Secrets (that we normally feed to the installer)
3) apply the generated Secrets to the live cluster

During install, some other pieces are responsible for creating the namespaces, but since we're basically doing some pre-upgrade steps, we need to create any namespaces for any Secrets that we generated (so that when the new component that will run in this new namespace is installed, it will have the credentials it needs immediately). The only times you would need to create a namespace is if there was a brand new component in this new version of OpenShift, and because we haven't yet upgraded, the new namespace isn't there yet. There is no fancy "create this secret and any namespace if necessary" command that I'm aware of, so you end up having to perform the two steps explicitly:
1) create the namespace to hold the Secret
2) create the Secret

Reviewer: And yes, we can leave the "verify the IAM roles created by querying AWS.." bit. it is harmless.

* To verify that the {product-title} secrets are created, list the files in the `_<path_to_ccoctl_output_dir>_/manifests` directory:
+
[source,terminal,subs="+quotes"]
----
$ ll __<path_to_ccoctl_output_dir>__/manifests
----
+
.Example output:
+
[source,terminal,subs="+quotes"]
----
total 24
-rw-------. 1 __<user>__ __<user>__ 161 Apr 13 11:42 cluster-authentication-02-config.yaml
-rw-------. 1 __<user>__ __<user>__ 379 Apr 13 11:59 openshift-cloud-credential-operator-cloud-credential-operator-iam-ro-creds-credentials.yaml
-rw-------. 1 __<user>__ __<user>__ 353 Apr 13 11:59 openshift-cluster-csi-drivers-ebs-cloud-credentials-credentials.yaml
-rw-------. 1 __<user>__ __<user>__ 355 Apr 13 11:59 openshift-image-registry-installer-cloud-credentials-credentials.yaml
-rw-------. 1 __<user>__ __<user>__ 339 Apr 13 11:59 openshift-ingress-operator-cloud-credentials-credentials.yaml
-rw-------. 1 __<user>__ __<user>__ 337 Apr 13 11:59 openshift-machine-api-aws-cloud-credentials-credentials.yaml

> ----

Reviewer: This section doesn't really make sense or add much value for the upgrade usecase.

// Module included in the following assemblies:
//
// * authentication/managing_cloud_provider_credentials/cco-mode-sts.adoc

[id="cco-ccoctl-upgrading_{context}"]
= Updating AWS resources with the Cloud Credential Operator utility

> The process for upgrading an {product-title} cluster configured for manual mode with STS is similar to installing on a cluster for which you create the AWS resources individually.

Reviewer 2: Suggestions to spell out STS on the first reference in each module. 
s/STS/AWS Secure Token Service (STS)

[NOTE]
====
By default, `ccoctl` creates objects in the directory in which the commands are run. To specify a directory, use the `--output-dir` flag. This procedure uses `_<path_to_ccoctl_output_dir>_` to refer to this location.

Some `ccoctl` commands make AWS API calls to create or modify AWS resources. To place JSON files on the local file system instead, use the `--dry-run` flag. These JSON files can be reviewed or modified and then applied with the AWS CLI tool using the `--cli-input-json` parameters.
====

.Prerequisites

* Obtain the {product-title} release image for the version that you are upgrading to.

* Extract and prepare the `ccoctl` binary from the release image.

.Procedure

. Extract the list of `CredentialsRequest` custom resources (CRs) from the {product-title} release image:
+
[source,terminal,subs="+quotes"]
----
$ oc adm release extract --credentials-requests --cloud=aws --to=__<path_to_directory_with_list_of_credentials_requests>__/credrequests quay.io/__<path_to>__/ocp-release:__<version>__
----

> . For each `CredentialsRequest` CR in the release image, ensure that a namespace exists in the `spec.secretRef.namespace` field.

Reviewer: Maybe just to make it abundantly clear why we need to do this step...
```suggestion
. For each `CredentialsRequest` CR in the release image, ensure that a namespace exists in the `spec.secretRef.namespace` field (this is where the generated Secrets holding the credentials configuration will be stored).
```

> . For each `CredentialsRequest` CR in the release image, ensure that a namespace exists in the `spec.secretRef.namespace` field. This field is where the generated secrets that hold the credentials configuration are stored.

Reviewer: Maybe way too pedantic, but this is all subtle enough as it is and I prefer to err on the side of over-explanation :)
```suggestion
. For each `CredentialsRequest` CR in the release image, ensure that a namespace exists in the cluster matching the text in the `spec.secretRef.namespace` field. This field is where the generated secrets that hold the credentials configuration are stored.
```

+
.Sample AWS `CredentialsRequest` object
[source,yaml]
----
apiVersion: cloudcredential.openshift.io/v1
kind: CredentialsRequest
metadata:
  name: cloud-credential-operator-iam-ro
  namespace: openshift-cloud-credential-operator
spec:
  secretRef:
    name: cloud-credential-operator-iam-ro-creds
    namespace: openshift-cloud-credential-operator <1>
  providerSpec:
    apiVersion: cloudcredential.openshift.io/v1
    kind: AWSProviderSpec
    statementEntries:
    - effect: Allow
      action:
      - iam:GetUser
      - iam:GetUserPolicy
      - iam:ListAccessKeys
      resource: "*"
----
+

> <1> This field must be populated.
> 
> . For any component that is missing this field, create a namespace in the `spec.secretRef.namespace` field:

Reviewer: I believe it isn't possible to not list the namespace. Or if it is technically possible, we would never allow that into the release image, because it would be ambiguous where the place the Secret with the credentials configuration.
```suggestion
<1> This field indicates the namespace which needs to exist to hold the generated Secret.

. For any CredentialsRequest where the cluster does not already have a namespace with the name specified in `spec.secretRef.namespace`, create the namespace:
```

----
$ ccoctl aws create-iam-roles --name __<name>__ --region=__<aws_region>__ --credentials-requests-dir=__<path_to_directory_with_list_of_credentials_requests>__/credrequests --identity-provider-arn arn:aws:iam::__<aws_account_id>__:oidc-provider/__<name>__-oidc.s3.__<aws_region>__.amazonaws.com
----
+
where:
+
** _<name>_ is the name used to tag any cloud resources that are created for tracking. For upgrades, use the same value that was used for the initial installation.
** _<aws_account_id>_ is the AWS account ID.
** _<aws_region>_ is the AWS region in which cloud resources will be created.
+
[NOTE]
====
For AWS environments that use alternative IAM API endpoints, such as GovCloud, you must also specify your region with the `--region` parameter.
====
+
For each `CredentialsRequest` object, `ccoctl` creates an IAM role with a trust policy that is tied to the specified OIDC identity provider, and a permissions policy as defined in each `CredentialsRequest` object from the {product-title} release image.

. Apply the secrets to your cluster:
+
[source,terminal,subs="+quotes"]
----
$ ls __<path_to_ccoctl_output_dir>__/manifests/*-credentials.yaml | xargs -I{} oc apply -f {}
----

> .Verification

Reviewer 2: Would it make sense to add an Additional Resources section to the assembly with a link to the AWS docs?
